[PATCH v5] api: tm: add rate limit mode in shapers

JannePeltonen opened the pull request:
Make it possible to do rate limiting instead of rate shaping in TM nodes
and queues. Introduce a mode parameter in shaper profile to select between
rate shaping and rate limiting. Add capability flags that tell if rate
limiting is supported.

Signed-off-by: Janne Peltonen <janne.peltonen@nokia.com>

nithind1988 commented:
Changes look fine to me. This way it should fit cleanly in Marvell CN10K platform.

JereLeppanen commented on include/odp/api/spec/traffic_mngr.h:
```diff
@@ -1047,13 +1045,44 @@ int odp_tm_drop_prec_marking(odp_tm_t           tm,
+	/** Rate limit traffic to the specified burst and rate by dropping
+	 *  excess packets. It is implementation dependent whether the limiter
+	 *  state update and packet drop occur immediately when packets are
+	 *  available from the source of when the downstream node or TM egress
```
"of when"? Should this be "or when"?

JannePeltonen replied:
Yes. Will fix.

JannePeltonen commented:
v3:
- Elaborated on the non-ideality of the rate limit feature in the API
- Added minimal validation test changes (not actual rate limit test)
- Fixed the typo

JannePeltonen commented:
v4: Tweaked the descriptive text in the API a bit based on offline comments.

nithind1988 commented:
Reviewed-By: Nithin Dabilpuram <ndabilpuram@marvell.com>
